lib/ex_oauth2_provider/device_grants/device_grants.ex

```diff
@@ -0,0 +1,94 @@
+defmodule ExOauth2Provider.DeviceGrants do
+  @moduledoc """
+  The boundary for the OauthDeviceGrants system.
+  """
+
+  import Ecto.Query
+  alias ExOauth2Provider.Mixin.{Expirable, Revocable}
+
+  alias ExOauth2Provider.{
+    Applications.Application,
+    DeviceGrants.DeviceGrant,
+    Config
+  }
+
+  defdelegate is_expired?(device_grant), to: Expirable
+  defdelegate revoke!(data, config \\ []), to: Revocable
+  defdelegate revoke(data, config \\ []), to: Revocable
+
+  def authorize(device_grant, resource_owner, config) do
+    params = %{resource_owner_id: resource_owner.id, user_code: nil}
+
+    device_grant
+    |> DeviceGrant.changeset(params, config)
+    |> get_repo(config).update()
+  end
+
+  @doc """
+  Creates a device grant.
+
+  ## Examples
+
+      iex> create_grant(application, attrs)
+      {:ok, %OauthDeviceGrant{}}
+
+      iex> create_grant(application, attrs)
+      {:error, %Ecto.Changeset{}}
+
+  """
+  @spec create_grant(Application.t(), map(), keyword()) ::
+          {:ok, DeviceGrant.t()} | {:error, term()}
+  def create_grant(application, attrs, config \\ []) do
+    config
+    |> device_grant_schema()
+    |> struct(application: application)
+    |> DeviceGrant.changeset(attrs, config)
+    |> get_repo(config).insert()
+  end
+
+  def delete_expired(config) do
+    schema = device_grant_schema(config)
+    lifespan = Config.authorization_code_expires_in(config)
+
+    from(
+      d in schema,
+      where: d.inserted_at <= ago(^lifespan, "second")
+    )
+    |> get_repo(config).delete_all()
+  end
+
+  def find_by_user_code(nil, _config), do: nil
+
+  # DeviceGrant | nil
+  def find_by_user_code(user_code, config) do
+    schema = device_grant_schema(config)
```

Comment:
Since you're always getting the schema and the repo why not get them both from one method? Also, some other suggestions to not make this so `assign a variable-y`

```elixir

config
|> schema_and_repo_from_config()
|> fetch_device_with_code(code)
|> execute_query()


def schema_and_repo_from_config(config) do
  {
    Config.device_grant(config),
    Config.repo(config)
  }
end

def fetch_device_with_code({schema, repo}, code) do
  schema
  |> where([entity], entity.user_code == ^code)
  |> preload([d], [:application])
  |> repo.one()
end

```

Reply:
I like! Working fast so haven't really looked back at anything like that but these kinda notes are always welcome.